cloud/linode/loadbalancers.go

```diff
@@ -309,16 +293,120 @@ func (l *loadbalancers) updateNodeBalancerFirewall(ctx context.Context, service
+			ID:   nb.ID,
+			Type: "nodebalancer",
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l *loadbalancers) updateFWwithACL(ctx context.Context, service *v1.Service, nb *linodego.NodeBalancer) error {
+	// See if a firewall is attached to the nodebalancer first.
+	firewalls, err := l.client.ListNodeBalancerFirewalls(ctx, nb.ID, &linodego.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	switch len(firewalls) {
+	case 0:
+		{
+			// need to create a fw and attach it to our nb
+			fwcreateOpts, err := l.createFirewallOptsForSvc(l.GetLoadBalancerName(ctx, "", service), l.getLoadBalancerTags(ctx, "", service), service)
+			if err != nil {
+				return err
+			}
+
+			fw, err := l.client.CreateFirewall(ctx, *fwcreateOpts)
+			if err != nil {
+				return err
+			}
+			// attach new firewall.
+			_, err = l.client.CreateFirewallDevice(ctx, fw.ID, linodego.FirewallDeviceCreateOptions{
 				ID:   nb.ID,
 				Type: "nodebalancer",
 			})
 			if err != nil {
 				return err
 			}
 		}
+	case 1:
+		{
+			// We do not want to get into the complexity of reconciling differences, might as well just pull what's in the svc annotation now and update the fw.
+			fwCreateOpts, err := l.createFirewallOptsForSvc(service.Name, []string{""}, service)
+			if err != nil {
+				return err
+			}
+			_, err = l.client.UpdateFirewallRules(ctx, firewalls[0].ID, fwCreateOpts.Rules)
+			if err != nil {
+				return err
+			}
+		}
+	default:
+		return errTooManyFirewalls
+	}
+	return nil
+}
+
+// updateNodeBalancerFirewall reconciles the firewall attached to the nodebalancer
+//
+// This function does the follwing
+//  1. If a firewallID annotation is present, it checks if the nodebalancer has a firewall attached, and if it matches the annotationID
+//     a. If the IDs match, nothing to do here.
+//     b. If they don't match, then the nb is removed from the old FW, and its attached to the new firewall
+//  2. If a firewallACL annotation is present,
+//     a. it checks if the nodebalancer has a firewall attached, if a fw exists, it updates rules
+//     b. if a fw does not exist, it creates one
+//  3. If neither of these annotations are present,
+//	  a. AND if no firewalls are attached to the nodebalancer, nothing to do.
+//	  b. if the NB has ONE firewall attached, remove it from nb, and clean up if nothing else is attached to it
+//	  c. If there are more than one fw attached to it, then its a problem, return an err
+// IF a user creates a fw ID externally, and then switches to using a ACL, the CCM will take over the fw that's attached to the nodebalancer.
+
+func (l *loadbalancers) updateNodeBalancerFirewall(ctx context.Context, service *v1.Service, nb *linodego.NodeBalancer) error {
+	// get the new firewall id from the annotation (if any).
+	_, fwIDExists := service.GetAnnotations()[annLinodeCloudFirewallID]
+	if fwIDExists { // If an ID exists, we ignore everything else and handle just that
```

Comment:
> we ignore everything else and handle just that

What about the case when _both_ annotations are present?

Reply:
added a line to clarify that